experimental/web/sample_webgpu/main.c

```diff
@@ -279,8 +293,192 @@ static iree_status_t parse_inputs_into_call(
+
+  // Clone into memory that isn't readonly.
+  // TODO(scotttodd): import as readonly? const void* -> void*?
+  void* data_cloned = NULL;
+  iree_host_size_t cloned_data_length =
+      iree_host_align(data_length, IREE_HAL_HEAP_BUFFER_ALIGNMENT);
+  if (iree_status_is_ok(status)) {
+    status = iree_allocator_malloc(iree_allocator_system(), cloned_data_length,
+                                   &data_cloned);
+  }
+  if (iree_status_is_ok(status)) {
+    memcpy(data_cloned, data_ptr, data_length);
+  }
+  wgpuBufferUnmap(buffer_handle);
+
+  // This looks a bit indirect, but |iree_hal_heap_buffer_wrap| isn't
+  // public while |iree_hal_allocator_import_buffer| is.
+  iree_hal_buffer_t* heap_buffer = NULL;
+  iree_hal_external_buffer_t external_buffer = {
+      .type = IREE_HAL_EXTERNAL_BUFFER_TYPE_HOST_ALLOCATION,
+      .flags = IREE_HAL_EXTERNAL_BUFFER_FLAG_NONE,
+      .size = cloned_data_length,
+      .handle = {{data_cloned}},
+  };
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_allocator_import_buffer(
+        userdata->heap_allocator,
+        (iree_hal_buffer_params_t){
+            .type = IREE_HAL_MEMORY_TYPE_HOST_LOCAL,
+            .usage = IREE_HAL_BUFFER_USAGE_MAPPING,
+            .min_alignment = IREE_HAL_HEAP_BUFFER_ALIGNMENT,
+        },
+        &external_buffer, iree_hal_buffer_release_callback_null(),
+        &heap_buffer);
+  }
+
+  // Copy the original buffer_view, backed by the cloned heap buffer instead.
+  iree_hal_buffer_view_t* buffer_view = NULL;
+  iree_hal_dim_t buffer_view_shape[10];
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_buffer_view_shape(userdata->buffer_view,
+                                        /*rank_capacity=*/10, buffer_view_shape,
+                                        /*out_shape_rank=*/NULL);
+  }
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_buffer_view_create(
+        heap_buffer, iree_hal_buffer_view_shape_rank(userdata->buffer_view),
+        buffer_view_shape,
+        iree_hal_buffer_view_element_type(userdata->buffer_view),
+        iree_hal_buffer_view_encoding_type(userdata->buffer_view),
+        iree_allocator_system(), &buffer_view);
+  }
+
+  if (iree_status_is_ok(status)) {
+    fprintf(stdout, "Call output:\n");
+    status = iree_hal_buffer_view_fprint(stdout, buffer_view,
+                                         /*max_element_count=*/4096,
+                                         iree_allocator_system());
+    fprintf(stdout, "\n");
+  }
+  iree_hal_buffer_view_release(buffer_view);
+  iree_hal_buffer_release(heap_buffer);
+
+  if (!iree_status_is_ok(status)) {
+    fprintf(stderr, "buffer_map_sync_callback error:\n");
+    iree_status_fprint(stderr, status);
+    iree_status_free(status);
+  }
+
+  iree_allocator_free(iree_allocator_system(), data_cloned);
+
+  iree_allocator_free(iree_allocator_system(), userdata);
+}
+
+static iree_status_t print_buffer_view(iree_hal_device_t* device,
```

Comment:
(fine for follow-up as this is still experimental, but I wouldn't want anyone taking ideas from this code as it is :)

you can rework this sequence to be over all readback buffers instead:
- iree_hal_device_queue_alloca to get the mappable_buffers, waiting on the invocation completion semaphore
- iree_hal_create_transfer_command_buffer to create the transfer command buffer (eventually we can cache this for reuse) with all transfers
- iree_hal_device_queue_execute to queue the transfer
- use iree_hal_fence_await on the wait fence -> iree_loop_wait_one and in the callback do your webgpu mappings

then with the async invocation API that returns a fence you can chain invocation->allocations->transfers all without any waiting (when queue-ordered allocations are implemented) - that's how we'd want real users to do it, and it'd make this code easier to follow too

(really when invoking in a loop we'd want to have in-place outputs so that we weren't allocating anything, but that's a separate can of worms :)

Reply:
Progress on this: https://github.com/openxla/iree/pull/13820 (still need to connect the Emscripten/JS Promise wait handle to `webgpu/semaphore_*` and use queue-ordered allocations though)